README.md
# Learning.mozilla.org

[![Build Status](https://travis-ci.org/mozilla/learning.mozilla.org.svg)](https://travis-ci.org/mozilla/learning.mozilla.org)
[![Shipping fast with zenHub](https://raw.githubusercontent.com/ZenHubIO/support/master/zenhub-badge.png)](https://zenhub.com)

  This is the source code for [learning.mozilla.org](https://learning.mozilla.org), using [learning.mofostaging.net](https://learning.mofostaging.net) as staging server.

**For a summary of the most recent changes to the project, please
see [`CHANGELOG.md`][changelog]. It's awesome.**

# Overview

This software consists of two major parts:

* A **static site generator** that creates a number of
  `index.html` files in various directories which can be viewed
  in any browser, including ones that don't support JavaScript.
* Client-side JavaScript code that **progressively enhances**
  the user experience based on browser capabilities.

It should be noted that, based on the
[product roadmap][roadmap], the static site generator
may eventually evolve into becoming a dynamic server.

# Get started

### Dependencies

To get a local version of the site running, you'll need to have [git](http://git-scm.com/) and [node](http://nodejs.org/) installed on your local machine.

### Clone

In order to contribute to this project, you'll need to **create your own fork** of it and make pull-requests against our master branch.

Clone from your own fork or from the original:

```
git clone git@github.com:mozilla/learning.mozilla.org.git
cd learning.mozilla.org
```

### Build and Develop

To start developing, all you need to do is run the following in the `learning.mozilla.org` directory you just created:

```
$> npm install
$> npm start
```

This will start a webserver for you at `http://localhost:8008`, and run a `watch` process so that your front-end assets will be regenerated as you make changes.

### A note about source maps

Because this project uses a variety of tools that transform the
original source code before delivering it to the browser, it is critical
for [source maps][] to work properly in order to ensure a pleasant
developer experience.

However, due to the unfortunate realities of software development, there
isn't a "one size fits all" solution to this. Please be sure to read the
**Environment Variables** section below and configure your environment
as needed to ensure that you're seeing useful line numbers for
whatever part of the codebase you need to change.

### Directory and naming conventions

JS
* `lib/`: Non-react modules, as well as entry-point modules like `main.jsx` and `routes.jsx`
* `components/`: Re-usable react components that can be used throughout the site
* `pages/`: React "page" components, i.e. top-level pages required by `lib/router`
* `hoc/`: React higher-order components

Less
* `less/`: This is where you should add styles. `common.less` is the entry-point.

Other
* `gulp/`: build tasks used for running the site, as well as testing
* `build/`: used as a staging area when the site gets built
* `test/`: For js tests, manual tests
* `img/`: For images
* `dist/`: Generated site assets goes here. **This folder is gitignored, do not edit files here.**

### Localization

Work in progress, but please read the documentation on how we are doing it for the site [here](./L10N.md).

### Test

Fully testing the code is accomplished by running `npm test`,
which exercises a number of different aspects of the
codebase described below.

#### Static Site Generation (Smoke Test)

This generates a full static site and
ensures that **no React warnings are raised**.

Individually running *only* the smoke test can be accomplished
via `npm run smoketest`.

#### Unit Tests

Unit tests are spread across two different testing
environments.

Both environments use the [mocha][] test runner and [should][]
for assertions.

##### Node Tests

These tests generally exercise the code of the static site generator
and are located in the `test` directory.

Each test file should end with `.test.js` and will be automatically
discovered by the test runner.

Individually running *only* the node unit tests can be accomplished
via `node_modules/.bin/mocha test/*.test.js`. For more options,
see the documentation for [mocha (1)][].

##### Browser Tests

These tests exercise the code that runs in the user's browser. They're
located in the `test/browser` directory.

Each test file should end with `.test.js` or `.test.jsx` and will be
automatically discovered by the test runner.

Individually running *only* the browser unit tests can be accomplished
by first running `npm start` and then visiting http://localhost:8008/test/
in your browser.

#### Manual Tests

For detailed information about manual testing, see the
[Manual Testing][] section of `CONTRIBUTING.md`.

## Generating A Static Site

Run `npm run build` to generate a static site in `dist/` that
doesn't *require* any client-side JavaScript. This static
site also uses `history.pushState` for navigation if the browser
supports it.

For reference, a recent static build of the site can be found at
[teach.mofostaging.net][].

See the **Environment Variables** section below for more details.

## Environment Variables

The following environment variables can be used to modify how the
software works.

   Name | Description
------------------|---------------------------------------------
`NODE_ENV` | set this to `production` to automatically minify code and remove various development-only affordances.
`SHOW_DEV_RIBBON` | set this to `on` to show the "dev version" ribbon even when `NODE_ENV` is set to production. This can be used on staging sites.
`WEBPACK_DEVTOOL` | determines the setting for the [`devtool`][] Webpack option. In development, it defaults to `eval`, while in production it defaults to `source-map`. For more details on the trade-offs between different options, see our [conversation on sourcemaps][sourcemaps].
`LESS_AUTOPREFIXER` | set this to `off` to disable the LESS autoprefixer and enable useful CSS source maps, which is a workaround for [#413][].
`ORIGIN` | is the domain name of which the site is hosted. This is used in situations where absolute URLs are required, such as generating a `sitemap.xml` file.
`MAPBOX_ACCESS_TOKEN` | is the [Mapbox][] access token to use for embedded maps in the website. Optional.
`MAPBOX_MAP_ID` | is the Mapbox map ID to use for embedded maps in the website. Optional.
`TEACH_API_URL` | is the origin of the [Teach API][] server. Defaults to `https://teach-api.herokuapp.com`.
`GA_ACCOUNT` | is the property ID of the Google Analytics account. E.g. `UA-123...`. It defaults to the property ID for the Teach site. Set it to `DISABLED` to disable Google Analytics entirely.
`GA_DEBUG` | When set to 'on' will output verbose info to the console about what data is being sent to Google Analytics.
`NEWSLETTER_MAILINGLIST_URL` | is the newsletter signup form destination. If undefined, newsletter signup is simulated for development purposes.
`PORT` | is the port to serve the experimental lightweight dynamic server on. Defaults to 8008.
`OPTIMIZELY_ID` | Optimizely Project ID (not a secret) e.g. '206878104'
`OPTIMIZELY_ACTIVE` | If set to 'yes' (String) the project will include Optimizely snippet in the page load
`MAKE_METADATA_URL` | The template source url to load users' Makes metadata. Username should be included in the URL as a variable. e.g., `https://{username}.makes.org/makes.json`
`WORDPRESS_DOMAIN` | The domain to the wordpress.com site this app retrieves data from. e.g., `helloworld.wordpress.com` (with no protocol),
`HPKP` | [HTTP Public Key Pinning](https://developer.mozilla.org/en/docs/Web/Security/Public_Key_Pinning) - requires at least two sha256s, space separated. i.e "asdf= qwer="
`supportedLocales` | enables specific locales, for example: `['en-US', 'id']` will only enable 2 locales. Default value is `*` which will enable all locales under [locales](./locales) directory.
`LOGIN_EMULATION` | determines whether or not the teach-api is used for login state management. defaults to false.
`LOGIN_EMULATION_LOGGEDOUT` | determines whether, when login emulation is on, the login state should be logged in or out. defaults to false.
`LOGIN_EMULATION_USERNAME` | determines the username that is used when login emulation is on and login status is not set to logged out. defaults to 'testuser'.

### Using Environment Variables in Local Development

Create a file called `.env` in the root of your project, and then add
all the environment variables you need into this using the following
format:

```
export GA_ACCOUNT='xxxxxxx'
export TEACH_API_URL='xxxxxxx'
```

### Feature Flags

These are environment variables that turn on experimental features that need to be tested in a staging environment but should not show up on a production deploy. These flags need to explicitly be set for the feature to show up.

The current list of feature flags is:

Name|If set, enables...
-|-
`ENABLE_BADGES` | rendering the `/badges` and `/badge/....` routes, and reveals the badge main page as a sidebar navigation option.

## Content Security Policies

If you run into CORS or CSP issues (blocked domains trying to GET or POST, loading external scripts like X-Ray Goggles, etc) then you probably want to make sure that the Content Security Policy is set to clear whatever it is you need cleared.

We handle this in `app.js`, in the `helmet.contentSecurityPolicy` block. Typically you'll care about the `connectSrc` block for remote content fetching, and the `scriptSrc` block for injecting/loading external scripts, although you might have CSP needs outside those two categories: all of the major CSP categories have their own section and should be readily identifiable.

## Lightweight Dynamic Server (Experimental)

The site can also be served through a lightweight server which
dynamically generates requested HTML files. This feature is currently
experimental. For more information on the rationale behind it, see
[#585][].

To run the server in development mode, use `npm run app`. The
server-side page rendering code will be dynamically updated as
relevant files are changed.

For production deployment, you will need to:

1. Set `NODE_ENV=production` and other relevant environment variables
   and then re-run `npm install`. This will build all static assets
   that won't change during production.

2. Run `node app.js`.

The server is also designed to be easily deployable on Heroku.

## References

* [Roadmap][roadmap]
* [Change Log][changelog]
* [Contribution Guidelines][contributing]

<!-- links -->

  [learning.mozilla.org]: https://learning.mozilla.org
  [`devtool`]: http://webpack.github.io/docs/configuration.html#devtool
  [sourcemaps]: https://github.com/mozilla/learning.mozilla.org/pull/147#discussion-diff-25879885
  [roadmap]: http://wiki.webmaker.org/Teach
  [mocha]: http://mochajs.org/
  [mocha (1)]: http://mochajs.org/#usage
  [should]: https://www.npmjs.com/package/should
  [teach.mofostaging.net]: https://teach.mofostaging.net/
  [Mapbox]: http://mapbox.com/
  [Teach API]: https://github.com/mozilla/teach-api
  [#413]: https://github.com/mozilla/learning.mozilla.org/issues/413
  [source maps]: http://blog.teamtreehouse.com/introduction-source-maps
  [changelog]: https://github.com/mozilla/learning.mozilla.org/blob/develop/CHANGELOG.md
  [contributing]: https://github.com/mozilla/learning.mozilla.org/blob/develop/CONTRIBUTING.md
  [Manual Testing]: https://github.com/mozilla/learning.mozilla.org/blob/develop/CONTRIBUTING.md#manual-testing
  [#585]: https://github.com/mozilla/learning.mozilla.org/issues/585


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file,
which uses the format described in
[keepachangelog.com](http://keepachangelog.com/). This project adheres
to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]

## [1.0.5][] - 2016-08-04
### Added
- New Clubs Application Form
- Enable Pontoon

## [1.0.4][] - 2016-08-02
### Added
- Clubs' Guides page containing resource guides used by Mozilla Clubs
- Localization - The application is now localizable and has been localized in English

### Changed
- Community call updated to July

## [1.0.3][] - 2016-07-07
### Added
- Added July 12th Curriculum Workshop details and updated archives.

********************************************************************
**And... things are already on production site but we forgot to mention them in the CHANGELOG**

### Added
- Visit https://learning.mozilla.org/codemoji to check out a browser-based
  learning tool / game that teaches everyday internet users about ciphers,
  the basic building blocks of encryption.
- Added Offline Icebreakers teaching activity page.

### Changed
- Improved newsletter signup form validation.
- Moved privacy policy agreement text above email input box in newsletter
  signup forms.

## [1.0.2][] - 2016-06-09
### Added
- Added June 16 Curriculum Workshop details and updated archives.
- Added June 29 MLN Community Call details and updated archives.

## [1.0.1][] - 2016-06-03
### Changed
- Updated references to teach.mozilla.org with learning.mozilla.org
- Updated Twitter link.

## [1.0.0][] - 2016-05-11
### Added
- Added Intermediate Web Lit II module.
- Badges related code has been merged into the codebase! The `ENABLE_BADGES`
  flag is currently off. Turning it on renders the `/badges` and
  `/badge/....` routes, and reveals the badge main page as a sidebar navigation
  option.

### Changed
- `npm start` now triggers the `npm run app` task

## [0.0.40][] - 2016-05-05
### Added
- Added May 10 Curriculum Workshop details and updated archives.
- Added May 25 MLN Community Call details and updated archives.

## [0.0.39][] - 2016-04-20
### Changed
- Fixed Web Literacy Map data since Synthesize, Share and Connect were
  assgined the wrong 21st century skills.
- Unified button, label, and link styles across the site.
- Fixed url-to-react-router-path rewriter so that hitting pages with hash
  and/or query string in the URL won't result in 404.

## [0.0.38][] - 2016-04-15
### Changed
- Fixed width scrollbars issue in vertical container for Firefox stable.
- Updated a few links.

### Added
- Updated Community Calls pages. This includes adding an archive page
  for the March 23 call.

## [0.0.37][] - 2016-04-01
### Changed
- Updated a few links on the site.

### Added
- Added Curriculum Workshop archived pages.
- Added PULL_REQUEST_TEMPLATE.md.

## [0.0.36][] - 2016-03-29
### Changed
- Added attribution on Web Lit Basics I module.
- Updated link to Mozilla's Web Literacy whitepaper.
- Converted url values for weblit map to lowercase and replaced spaces
  with dashes.
- The site now has a real server! Pages on https://teach.mozilla.org are
  served by the server and are no longer static files live on S3.
- Removed S3 related gulp tasks and files.
- Both staging the production instances now live on Heroku instead. Auto-
  deployment to staging and production is triggered by push to `develop`
  and `master` branches respectively.

## [0.0.35][] - 2016-03-24
### Changed
- Edited v0.0.34 release changelog message - production site hasn't been switched to
  use a real server yet.

### Added
- Woohooo! It's live now! Check out the interactive web literacy map on
  https://teach.mozilla.org/web-literacy and 21st Century Skills on
  https://teach.mozilla.org/web-literacy/skills

## [0.0.34][] - 2016-03-22
### Changed
- The site now has a real server! ~~Pages on https://teach.mozilla.org are
  served by the server and are no longer static files live on S3.~~ (edited on 2016-03-24)
- Turned off auto push to S3 when the `develop` branch is updated.
- Updated Mouse logo.
- Removed Private Eye page.

### Added
- Added Community Call page.
- Added 404 messages server side.

## [0.0.33][] - 2016-03-10
### Changed
- Removed Creative Commons GIF Exchange module and CSS Story Cards module.
- Minor text updates to Curriculum Workshop page.

### Added
- Added two new modules to /activities, the Mouse module and Intermediate
  Web Literacy module.

## [0.0.32][] - 2016-03-02
### Changed
- Updated the donate link.
- Replaced Pledge to Teach CTA with newsletter signup.
- Updated newsletter subscription url.

### Added
- Added Curriculum Workshop page.

## [0.0.31][] - 2016-02-16
### Changed
- The build system for the site got an overhaul.
- Links to the Goggles were updated to point to its new home.
- The "Creating Commons GIF exchange" activity got a new link.
- The "CSS Story Cards" activity got a new link.

### Added
- Added Let's Encrypt campaign material.

## [0.0.30][] - 2016-01-14
### Changed
- The Teach Like Mozilla page has been removed.
- Nav links on the sidebar have been reorganized.
- More UI tweaks! We've updated many icons on the site to make sure they
  share the same styling.

### Added
- Added Leadership Opportunities page.
- Added Parapara teaching kit.
- Added link to the Online Tracking module.

## [0.0.29][] - 2015-12-22
### Changed
- Updated dependencies including uplifting React to v0.14 and other packages.
- Started UI unification work.
- Updated page banners across the site.

### Added
- Added Webmaker activity kit.
- Added Community page.


## [0.0.28][] - 2015-12-04
### Changed
- Replaced the sidebar with a cleaner design.
- Fixed broken links to Discourse.

## [0.0.27][] - 2015-12-03
### Added
- Added a Made with Code activity kit.
- Added Web Lit Basics II activity kit.
- Updated README.md to include platform specific setup instructions.
- "Help us translate" CTA has been added to the Web Literacy Basics page.

### Changed

app.js
var path = require('path');
var fs = require('fs');
var express = require('express');
var helmet = require('helmet');
var url = require('url');

var React = require('react');
var ReactRouter = require('react-router');
var Router = ReactRouter.Router;
var match = ReactRouter.match;
var routington = require('routington');

var indexStaticWatcher = require('./lib/build/index-static-watcher').create();
var PORT = process.env.PORT || 8008;
var PRODUCTION = (process.env.NODE_ENV === 'production');
var DIST_DIR = path.join(__dirname, 'dist');
var CODEMOJI_URL = process.env.CODEMOJI_URL || "https://codemoji.mofostaging.net";
var localize = require('mofo-localize');

var WpPageChecker = require('./lib/wp-page-checker');

var habitat = require('habitat');
habitat.load('.env');

var indexStatic;
var router;
var matcher;
var app = express();
var locale = "";
var locales = require('./dist/locales.json');

app.disable('x-powered-by');

var notFoundHTML = [
  '<!doctype html>',
  '<html><head>',
  '<meta charset="utf-8">',
  '<title>404 - Page not found</title>',
  '</head><body>',
  '<p>404 - Page not found</p>',
  '</body></html>'
].join('');

var urlToRoutePath = function(loc) {
  // For router-resolution, we don't want hashes...
  var pos = loc.indexOf('#');
  if (pos > -1) { loc = loc.substring(0,pos); }
  // And we don't want query strings, either...
  pos = loc.indexOf('?');
  if (pos > -1) { loc = loc.substring(0,pos); }
  // If this is not the site root, we need to remove the leading slash.
  if (loc !== '/') {
    loc = loc.replace(/^\//, '').replace(/\/$/, '');
  }
  return loc;
};

var startProdApp = function() {
  console.log([
    'Production mode enabled. Note that "npm install" is assumed to',
    'have recently been run with NODE_ENV="production". If this is not',
    'the case, some or all static assets may be out of date.'
  ].join('\n'));
  indexStaticWatcher.build(function(err, newIndexStatic) {
    if (err) {
      throw err;
    }

    console.log('Built server-side bundle.');
    updateIndexStatic(newIndexStatic);
    app.listen(PORT, function() {
      console.log('Listening on port', PORT);
    });
  });
};

var updateIndexStatic = function(newIndexStatic) {
  indexStatic = newIndexStatic;
  router = indexStatic ? React.createElement(Router, {routes: indexStatic.routes}) : null;
};

// make sure the dir we'll be using for static hosting exists.
if (!fs.existsSync(DIST_DIR)) {
  fs.mkdirSync(DIST_DIR);
}

/**
 * Security Headers
 */
 var securityHeaders = {
  directives: {
    defaultSrc: [
      'www.youtube.com',
      'https://public.etherpad-mozilla.org'
    ],
    scriptSrc: [
      '\'self\'',
      '\'unsafe-inline\'',
      '\'unsafe-eval\'',
      'data:',
      'www.google-analytics.com',
      'cdn.optimizely.com',
      'https://www.google.com',
      'https://s.ytimg.com',
      'https://www.mozilla.org'
    ],
    fontSrc: [
      '\'self\'',
      'fonts.googleapis.com',
      'fonts.gstatic.com'
    ],
    styleSrc: [
      '\'self\'',
      '\'unsafe-inline\'',
      'https://www.google.com',
      'fonts.googleapis.com',
      'https://api.tiles.mapbox.com',
      'https://s.ytimg.com'
    ],
    imgSrc: [
      '\'self\'',
      '\'unsafe-inline\'',
      '*'
    ],
    connectSrc: [
      '\'self\'',
      'https://www.google.com',
      '*.tiles.mapbox.com',
      '*.log.optimizely.com',
      '*.github.io',
      '*.mywebmaker.org',
      '*.makes.org',
      'bitly.mofoprod.net',
      process.env.TEACH_API_URL || 'https://teach-api-staging.herokuapp.com',
      url.parse(process.env.NEWSLETTER_MAILINGLIST_URL || 'https://basket-dev.allizom.org').hostname
    ]
  },
  reportOnly: false,
  browserSniff: false
};

if (process.env.ENABLE_PONTOON) {
  securityHeaders.directives.defaultSrc.push('https://pontoon.mozilla.org', 'https://mozilla-pontoon-staging.herokuapp.com');
  securityHeaders.directives.scriptSrc.push('https://pontoon.mozilla.org', 'https://mozilla-pontoon-staging.herokuapp.com');
  securityHeaders.directives.fontSrc.push('https://pontoon.mozilla.org', 'https://mozilla-pontoon-staging.herokuapp.com');
  securityHeaders.directives.styleSrc.push('https://pontoon.mozilla.org', 'https://mozilla-pontoon-staging.herokuapp.com');
  securityHeaders.directives.connectSrc.push('https://pontoon.mozilla.org', 'https://mozilla-pontoon-staging.herokuapp.com');
  securityHeaders.directives['frame-ancestors'] = ['https://pontoon.mozilla.org', 'https://mozilla-pontoon-staging.herokuapp.com'];
}
app.use(helmet.contentSecurityPolicy(securityHeaders));

app.use(helmet.xssFilter({
  setOnOldIE: true
}));

app.use(helmet.hsts({
  maxAge: 1000 * 60 * 60 * 24 * 90
}));

app.use(helmet.ieNoOpen());

app.use(helmet.noSniff());

if (process.env.HPKP) {
  app.use(helmet.hpkp({
    maxAge: 1000 * 60 * 60 * 24 * 90,
    sha256s: process.env.HPKP.split(' '),
    setIf: function (req, res) {
      return req.secure;
    }
  }));
}


/**
 * Wait for the router to come online.
 */
app.use(function(req, res, next) {
  if (router) {
    return next();
  }
  res.send('Please wait while the server-side bundle regenerates.');
});

/**
 * If we have a router, check if we're dealing with a redirect.
 */
app.use(function(req, res, next) {
  var url = urlToRoutePath(req.path);
  if (!indexStatic.REDIRECTS[url]) {
    return next();
  }
  res.redirect('/' + indexStatic.REDIRECTS[url] + '/');
});

/**
 * If it's not a redirect, is it a component page?
 */
app.use(function(req, res, next) {
  var routes = indexStatic.routes;
  var location = urlToRoutePath(req.url);
  var urls = indexStatic.URLS;
  var lang;

  if (!matcher) {
    matcher = routington();
    urls.forEach(function(route) { matcher.define(route); });
  }

  match({ routes: routes, location: location}, function resolveRoute(err, redirect, props) {
    // is this even a component?
    if ( !props ) {
      return next();
    }

    // if this belongs to one of the predefined urls, let's generate its associated page
    if ( matcher.match(location) ) {
      var search = url.parse(req.url).search || "";

      locale = localize.parseLocale(req.headers["accept-language"], location, locales).locale;
      if (location === "/") {
        res.redirect(302, location + locale + search);
        return;
      }

      lang = location.split('/')[0];
      return renderComponentPage(location, res, lang);
    }

    // if neither of those, this is wordpress content
    else {
      // check to see a page with this slug exists on the WordPress site
      WpPageChecker(location, function(error, wpContent) {
        if ( error ) {
          return next();
        }
        // WP page exists, let's load the WordPress content through a component page
        return renderComponentPage(location,res);
      });
    }
  });
});

function renderComponentPage(location, res, locale) {
  indexStatic.generate(location, {locale: locale}, function(err, location, title, html) {
    if (err) {
      next(err);
    }
    res.type('html').send(html);
  });
}

/**
 * codemoji redirect - https://github.com/mozilla/codemoji
 */
app.use('/codemoji', function(req, res) {
  var location = url.parse(CODEMOJI_URL);
  location.pathname = req.path;
  location.query = req.query;
  res.redirect(307, url.format(location));
});

/**
 * Is this a static asset?
 */
app.use(express.static(DIST_DIR));


/**
* Maybe it is a route, but needs the localized path
*/
app.use(function(req, res, next) {
  var location = url.parse(req.url).pathname;
  var search = url.parse(req.url).search || "";
  // Get a valid locale from the path and header
  var parsed = localize.parseLocale(req.headers["accept-language"], location, locales);
  var parsedLocale = parsed.locale;
  var parsedRedirect = parsed.redirect;
  // See if we should redirect.
  if (parsedRedirect) {
    res.redirect(307, "/" + parsedLocale + parsedRedirect + search);
  } else {
    next();
  }
});

app.use(function(req, res, next) {
  res.status(404).send(notFoundHTML);
});

app.DIST_DIR = DIST_DIR;
app.updateIndexStatic = updateIndexStatic;
module.exports = app;

if (!module.parent) {
  console.log('Initializing server.');

  if (PRODUCTION) {
    startProdApp();
  } else {
    console.log([
      'This server can only be run as a script when NODE_ENV="production".',
      'To run it in development mode, please use "npm run app".'
    ].join('\n'));
    process.exit(1);
  }
}
